ocean_lib/ocean/v4/ocean_assets.py

```diff
@@ -0,0 +1,408 @@
+            logger.info(
+                f"Successfully created data token with address "
+                f"{erc721_token.address} for new dataset asset."
+            )
+            erc721_address = erc721_token.address
+        else:
+            # verify nft address
+            if not erc721_factory.verify_nft(erc721_address):
+                raise ContractNotFound(
+                    f"NFT address {erc721_address} is not found in the ERC721Factory events."
+                )
+
+        assert erc721_address, "nft_address is required for publishing a dataset asset."
+        erc721_token = ERC721Token(self._web3, erc721_address)
+
+        # Create a DDO object
+        asset = V4Asset()
+
+        # Generating the did and adding to the ddo.
+        did = f"did:op:{create_checksum(erc721_token.address + str(self._web3.eth.chain_id))}"
+        asset.did = did
+        # Check if it's already registered first!
+        if self._get_aquarius().ddo_exists(did):
+            raise AquariusError(
+                f"Asset id {did} is already registered to another asset."
+            )
+        asset.chain_id = self._web3.eth.chain_id
+        asset.metadata = metadata
+
+        asset.credentials = credentials if credentials else build_credentials_dict()
+
+        erc20_addresses = []
+        services = services or []
+        deployed_erc20_tokens = deployed_erc20_tokens or []
+        if not deployed_erc20_tokens:
+            for erc20_token_data in erc20_tokens_data:
+                erc20_addresses.append(
+                    self.deploy_datatoken(
+                        erc721_factory=erc721_factory,
+                        erc721_token=erc721_token,
+                        erc20_data=erc20_token_data,
+                        from_wallet=publisher_wallet,
+                    )
+                )
+            if not services:
+                for erc20_address in erc20_addresses:
+                    services = self._add_defaults(
+                        services, erc20_address, encrypted_files, provider_uri
+                    )
+            for erc20_token_address in erc20_addresses:
+                deployed_erc20_tokens.append(
+                    ERC20Token(self._web3, erc20_token_address)
+                )
+
+            data_tokens = self.build_data_tokens_list(
+                services=services, deployed_erc20_tokens=deployed_erc20_tokens
+            )
+        else:
+            if not services:
+                for erc20_token in deployed_erc20_tokens:
+                    services = self._add_defaults(
+                        services, erc20_token.address, encrypted_files, provider_uri
+                    )
+
+            data_tokens = self.build_data_tokens_list(
+                services=services, deployed_erc20_tokens=deployed_erc20_tokens
+            )
+
+        asset.datatokens = data_tokens
+
+        for service in services:
+            asset.add_service(service)
+
+        nft = {
```

Comment:
it is not ocean.py's job to build the nft key. That is added by aquarius AFTER the MetadataCreated event is processed. It has nothing to do with validation and you shouldn't build it.